x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportPutFollowAction.java

```diff
@@ -131,98 +127,95 @@ private void createFollowerIndex(
+        initiateRestore(restoreRequest, restoreCompleteHandler);
+    }
 
-                RoutingTable.Builder routingTableBuilder = RoutingTable.builder(updatedState.routingTable())
-                        .addAsNew(updatedState.metaData().index(request.getFollowRequest().getFollowerIndex()));
-                updatedState = allocationService.reroute(
-                        ClusterState.builder(updatedState).routingTable(routingTableBuilder.build()).build(),
-                        "follow index [" + request.getFollowRequest().getFollowerIndex() + "] created");
+    private void initiateRestore(RestoreService.RestoreRequest restoreRequest, ActionListener<RestoreSnapshotResponse> listener) {
+        threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(new AbstractRunnable() {
+            @Override
+            public void onFailure(Exception e) {
+                listener.onFailure(e);
+            }
 
-                logger.info("[{}] creating index, cause [ccr_create_and_follow], shards [{}]/[{}]",
-                        followIndex, followIMD.getNumberOfShards(), followIMD.getNumberOfReplicas());
+            @Override
+            protected void doRun() {
+                restoreService.restoreSnapshot(restoreRequest, new ActionListener<RestoreService.RestoreCompletionResponse>() {
+                    @Override
+                    public void onResponse(RestoreService.RestoreCompletionResponse restoreCompletionResponse) {
+                        RestoreClusterStateListener.createAndRegisterListener(clusterService, restoreCompletionResponse, listener);
+                    }
 
-                return updatedState;
+                    @Override
+                    public void onFailure(Exception e) {
+                        listener.onFailure(e);
+                    }
+                });
             }
         });
+
     }
 
     private void initiateFollowing(
-            final PutFollowAction.Request request,
-            final ActionListener<PutFollowAction.Response> listener) {
+        final Client client,
+        final PutFollowAction.Request request,
+        final ActionListener<PutFollowAction.Response> listener) {
         activeShardsObserver.waitForActiveShards(new String[]{request.getFollowRequest().getFollowerIndex()},
-                ActiveShardCount.DEFAULT, request.timeout(), result -> {
-                    if (result) {
-                        client.execute(ResumeFollowAction.INSTANCE, request.getFollowRequest(), ActionListener.wrap(
-                                r -> listener.onResponse(new PutFollowAction.Response(true, true, r.isAcknowledged())),
-                                listener::onFailure
-                        ));
-                    } else {
-                        listener.onResponse(new PutFollowAction.Response(true, false, false));
-                    }
-                }, listener::onFailure);
+            ActiveShardCount.DEFAULT, request.timeout(), result -> {
```

Comment:
I'm not sure whether we need the waitForActiveShards logic as RestoreService will only return if the restore is completed (or failed)